--- pkg/mod/github.com/fatedier/frp@v0.39.0/cmd/frpc/sub/root.go	2022-02-07 09:18:00.000000000 +0800
+++ patches/root.go	2022-02-07 10:12:45.000000000 +0800
@@ -102,13 +102,13 @@
 		if showVersion {
 			fmt.Println(version.Full())
 			return nil
 		}

 		// Do not show command usage here.
-		err := runClient(cfgFile)
+		err := RunClient(cfgFile)
 		if err != nil {
 			fmt.Println(err)
 			os.Exit(1)
 		}
 		return nil
 	},
@@ -161,13 +161,13 @@
 		err = fmt.Errorf("Parse config error: %v", err)
 		return
 	}
 	return
 }

-func runClient(cfgFilePath string) error {
+func RunClient(cfgFilePath string) error {
 	cfg, pxyCfgs, visitorCfgs, err := config.ParseClientConfig(cfgFilePath)
 	if err != nil {
 		return err
 	}
 	return startService(cfg, pxyCfgs, visitorCfgs, cfgFilePath)
 }
