--- pkg/mod/github.com/fatedier/frp@v0.43.0/client/service.go	2022-04-22 14:59:20.000000000 +0800
+++ patches/service.go	2022-04-22 15:10:29.000000000 +0800
@@ -39,12 +39,13 @@
 	"github.com/fatedier/frp/pkg/util/version"
 	"github.com/fatedier/frp/pkg/util/xlog"
 	"github.com/fatedier/golib/crypto"
 	libdial "github.com/fatedier/golib/net/dial"

 	fmux "github.com/hashicorp/yamux"
+	frpmgr "github.com/koho/frpmgr/pkg/util"
 )

 func init() {
 	crypto.DefaultSalt = "frp"
 	rand.Seed(time.Now().UnixNano())
 }
@@ -271,14 +272,20 @@
 		libdial.WithProxyAuth(auth),
 		libdial.WithTLSConfig(tlsConfig),
 		libdial.WithAfterHook(libdial.AfterHook{
 			Hook: frpNet.DialHookCustomTLSHeadByte(tlsConfig != nil, svr.cfg.DisableCustomTLSFirstByte),
 		}),
 	)
+	var serverAddr = svr.cfg.ServerAddr
+	if svr.cfg.DNSServer != "" {
+		if serverAddr, err = frpmgr.LookupIP(svr.cfg.ServerAddr, svr.cfg.DNSServer); err != nil {
+			return nil, nil, err
+		}
+	}
 	conn, err = libdial.Dial(
-		net.JoinHostPort(svr.cfg.ServerAddr, strconv.Itoa(svr.cfg.ServerPort)),
+		net.JoinHostPort(serverAddr, strconv.Itoa(svr.cfg.ServerPort)),
 		dialOptions...,
 	)
 	if err != nil {
 		return
 	}
